src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+                    return Task.CompletedTask;
+                }
+                else if (errorCode != ERROR_IO_PENDING)
+                {
+                    if (!_fileHandle.IsClosed && CanSeek)  // Update Position - It could be anywhere.
+                    {
+                        SeekCore(_fileHandle, 0, SeekOrigin.Current);
+                    }
+
+                    completionSource.ReleaseNativeResource();
+
+                    if (errorCode == ERROR_HANDLE_EOF)
+                    {
+                        throw Error.GetEndOfFile();
+                    }
+                    else
+                    {
+                        throw Win32Marshal.GetExceptionForWin32Error(errorCode, _path);
+                    }
+                }
+                else if (cancellationToken.CanBeCanceled) // ERROR_IO_PENDING
+                {
+                    // Only once the IO is pending do we register for cancellation
+                    completionSource.RegisterForCancellation(cancellationToken);
+                }
+            }
+            else
+            {
+                // Due to a workaround for a race condition in NT's ReadFile &
+                // WriteFile routines, we will always be returning 0 from WriteFileNative
+                // when we do async IO instead of the number of bytes written,
+                // irregardless of whether the operation completed
+                // synchronously or asynchronously.  We absolutely must not
+                // set asyncResult._numBytes here, since will never have correct
+                // results.
+            }
+
+            return completionSource.Task;
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            ValidateCopyToArguments(destination, bufferSize);
+
+            // Fail if the file was closed
+            if (_fileHandle.IsClosed)
+            {
+                throw Error.GetFileNotOpen();
+            }
+            if (!CanRead)
+            {
+                throw Error.GetReadNotSupported();
+            }
+
+            // Bail early for cancellation if cancellation has been requested
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            return AsyncModeCopyToAsync(destination, bufferSize, cancellationToken);
+        }
+
+        private async Task AsyncModeCopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!_fileHandle.IsClosed, "!_handle.IsClosed");
+            Debug.Assert(CanRead, "_parent.CanRead");
+
+            bool canSeek = CanSeek;
+            if (canSeek)
+            {
+                VerifyOSHandlePosition();
+            }
+
+            try
+            {
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task
```

Comment:
Why did you suppress this?

Reply:
The code that I've moved to `FileStreamHelpers.AsyncModeCopyToAsync` was previously part of this method (inlined) so I assumed that I don't need `.ConfigureAwait(false)` and my micro-optimizations obsession told me to do it.

Should I remove the `#pragma` and add  `.ConfigureAwait(false)` ?

Reply:
> Should I remove the #pragma and add .ConfigureAwait(false) ?

Yes

> my micro-optimizations obsession told me to do it

The code has become more expensive by now having an extra async method awaiting another async method.